### PR TITLE
Adding Gen types for ScriptContexts

### PR DESCRIPTION
**Pull request by jhodgdev**

### CODE REVIEW

**Comment by jhodgdev** [comment]
Will add docstrings before submitting for review.

**Comment by jhodgdev** [comment]
As I alluded to in the sprint, I am worried what I figured would be a 'quick thing useful for upstreaming' ran away from me a bit, so I don't know what in here is worth keeping but (I am sorry to say) there's a lot! Appreciate your review @delicious-lemon and/or @Geometer1729 .

File: src/Apropos/Gen/Api.hs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+builtinData = do
+  d <- data'
+  return $ BuiltinData d
+
+-- | `Gen` for Plutus `Data`.
+data' :: Gen Data
+data' = do
+  let mapGen :: Gen [(Data, Data)]
+      mapGen = list (linear 0 20) $ pair data' data'
+  n <- integer $ linear minBound maxBound
```
Comment (Geometer1729) [inline src/Apropos/Gen/Api.hs]: I think generating these upfront and using element is worse for performance than using choice and generating them in their list entries.

File: src/Apropos/Gen/Value.hs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+-- | Generator for a Plutus `Value`.
+value :: Gen Value
+value = do
+  let tnvMap = Gen.map (linear 1 5) tokenName $ integer (linear 1 maxBound)
+  v <- Gen.map (linear 1 5) currencySymbol tnvMap
+  return $ Value v
+
+-- | Generator for a Plutus `CurrencySymbol`.
+currencySymbol :: Gen CurrencySymbol
```
Comment (Geometer1729) [inline src/Apropos/Gen/Value.hs:56]: `currencySymbol` and `tokenName` should probably be able to generate `""` for Ada.

**Review by Geometer1729 — CHANGES_REQUESTED** [review]

File: src/Apropos/Gen/Api.hs (hunk 1 above)
Comment (jhodgdev) [reply src/Apropos/Gen/Api.hs]: Interesting! ELI5?

File: src/Apropos/Gen/Value.hs (hunk 2 above)
Comment (jhodgdev) [reply src/Apropos/Gen/Value.hs:56]: I think it might be better to include some function `ada :: Range -> Gen Value` and another `valueWithAda :: Gen Value` which is the same as `value :: Gen Value` but includes some amount of Ada. Thoughts?

File: src/Apropos/Gen/Api.hs (hunk 1 above)
Comment (delicious-lemon) [reply src/Apropos/Gen/Api.hs]: :smiley: ELI5 attempt: Generating a list of things and choosing one of them is less efficient than choosing from a list of generators.

```Haskell
i <- genInt
s <- genString
element [I i, S s]
```
generates two things, one of which gets thrown away, whereas

```Haskell
choice [I <$> genInt, S <$> genString]
```
only generates one thing.

File: src/Apropos/Gen/Value.hs (hunk 2 above)
Comment (delicious-lemon) [reply src/Apropos/Gen/Value.hs:56]: It would be nice to have a model for `SingletonValue` that is parameterisable by tokens in a protocol. e.g.

```Haskell
data SingletonValueProp protocolToken =
    IsAda
  | IsProtocolToken protocolToken
  | IsOtherToken
```
Then this can be abstracted into a generator for `MultiValue` that is paramaterised by these value properties.

```Haskell
data MultiValue protocolToken =
  MultiValue (SingletonValue protocolToken)
```

File: src/Apropos/Gen/Api.hs (hunk 1 above)
Comment (srid) [reply src/Apropos/Gen/Api.hs:40]: The same would apply for sum types, like [`scriptPurpose`](https://github.com/mlabs-haskell/apropos-tx/pull/41/files#diff-3bdee9ceafcb87ab1d7e8d77038ba042c553ab4204647170acbe115fcbf85a83R107-R108).

Comment (srid) [reply src/Apropos/Gen/Api.hs]: The same would apply for sum types, like [`scriptPurpose`](https://github.com/mlabs-haskell/apropos-tx/pull/41/files#diff-3bdee9ceafcb87ab1d7e8d77038ba042c553ab4204647170acbe115fcbf85a83R107-R108).

**Comment by srid** [comment]
@jhodgdev @delicious-lemon Any thoughts on making the bulk of `Apropos.Gen.*` in this PR as a library of its own, so it can be used in other projects, like Plutarch? Or do you envision these generators to be tightly coupled with apropos properties over time?

**Comment by jhodgdev** [comment]
@srid I had always envisaged this work as only having utility for Apropos. You're more than welcome to use whatever you want from it though. What're your thoughts @delicious-lemon?

File: src/Apropos/Gen/Value.hs (hunk 2 above)
Comment (jhodgdev) [reply src/Apropos/Gen/Value.hs:56]: Sorry for re-visiting this after so long but I am not sure what you mean @delicious-lemon. I have added the described `ada` and `valueWithAda` functions already. Could we move this into a separate issue?

**Comment by jhodgdev** [comment]
So, I have been using this branch on Agora and am getting `CekEvaluationError`s. Is that a problem stemming from the `ScriptContexts` I am using being non-sensical, or...? @delicious-lemon @Geometer1729
